package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -65,6 +78,187 @@ def unique_int_1d(np.ndarray[np.int64_t, ndim=1] values):
+    """
+    cdef intset agset, refpoints, todo, done
+    cdef int i, nloops, ref, atom, other
+    cdef intmap bonding
+    cdef int[:, :] bonds
+    cdef float[:, :] oldpos, newpos
+    cdef bint ortho
+    cdef float[:] box
+    cdef float tri_box[3][3]
+    cdef float inverse_box[3]
+    cdef double vec[3]
+
+    if reference_atom is None:
+        ref = atomgroup[0].index
+    else:
+        # Sanity check
+        if not reference_atom in atomgroup:
+            raise ValueError("Reference atom not in atomgroup")
+        ref = reference_atom.index
+
+    box = atomgroup.dimensions
+
+    for i in range(3):
+        if box[i] == 0.0:
+            raise ValueError("One or more dimensions was zero.  "
+                             "You can set dimensions using 'atomgroup.dimensions='")
+
+    ortho = True
+    for i in range(3):
+        if box[i + 3] != 90.0:
+            ortho = False
+
+    if ortho:
+        for i in range(3):
+            inverse_box[i] = 1.0 / box[i]
+    else:
+        from .mdamath import triclinic_vectors
+        tri_box = triclinic_vectors(box)
+
+    # set of indices in AtomGroup
+    agset = intset()
+    for i in atomgroup.indices.astype(np.int32):
+        agset.insert(i)
+    # C++ dict of bonds
+    try:
+        bonds = atomgroup.bonds.to_indices()
+    except (AttributeError, NoDataError):
+        raise NoDataError("The atomgroup is required to have bonds")
+    for i in range(bonds.shape[0]):
+        atom = bonds[i, 0]
+        other = bonds[i, 1]
+        # only add bonds if both atoms are in atoms set
+        if agset.count(atom) and agset.count(other):
+            bonding[atom].insert(other)
+            bonding[other].insert(atom)
+
+    oldpos = atomgroup.positions
+    newpos = np.zeros((oldpos.shape[0], 3), dtype=np.float32)
+
+    done = intset()  # Who have I already done?
+    refpoints = intset()  # Who is safe to use as reference point?
+    # initially we have one starting atom whose position we trust
+    refpoints.insert(ref)
+    for i in range(3):
+        newpos[ref, i] = oldpos[ref, i]
+
+    nloops = 0
+    while refpoints.size() < agset.size() and nloops < agset.size():
+        nloops += 1
+
+        # We want to iterate over atoms that are good to use as reference
+        # points, but haven't been done yet.
+        todo = difference(refpoints, done)
+        for atom in todo:
+            for other in bonding[atom]:
+                # If other is already a refpoint, leave alone
+                if refpoints.count(other):
+                    continue
+                # Draw vector from atom to other
+                for i in range(3):
+                    vec[i] = oldpos[other, i] - oldpos[atom, i]
+                # Apply periodic boundary conditions to this vector
+                if ortho:
+                    minimum_image(&vec[0], &box[0], &inverse_box[0])
+                else:
+                    minimum_image_triclinic(&vec[0], <coordinate*>&tri_box[0])
```

Comment:
Couldn't `tri_box` be casted only once before the loops?

Reply:
does casting take any time? I thought it was more an instruction to the compiler rather than a copy of data?

Reply:
I do not expect any cost for the cast; though, I do not see why tri_box could not be directly the right type. Nevertheless, it is really not a big deal.